galter_subjects_utils/converter.py

```diff
@@ -21,39 +28,206 @@ def __init__(self, topics, qualifiers=None):
         :type qualifiers: iterable[dict], optional
         """
         self.topics = topics
-        self._qualifier_map = {
-            q.get("QA"): q for q in qualifiers
-        } if qualifiers else {}
-
-    def generate_id(self, identifier):
-        """Generate URI id."""
-        return f"https://id.nlm.nih.gov/mesh/{identifier}"
+        self.qualifiers_mapping = qualifiers_mapping
 
     def __iter__(self):
         """Iterate over converted entries."""
         for topic in self.topics:
             yield {
-                "id": self.generate_id(topic['UI']),
+                "id": generate_id(topic['UI']),
                 "scheme": "MeSH",
                 "subject": topic['MH']
             }
 
-            if not self._qualifier_map:
+            if not self.qualifiers_mapping:
                 continue
 
             for qid in topic.get("AQ", []):
-                qualifier = self._qualifier_map.get(qid)
+                qualifier = self.qualifiers_mapping.get(qid)
 
                 if not qualifier:
                     continue
 
                 yield {
-                    "id": self.generate_id(topic['UI'] + qualifier['UI']),
+                    "id": generate_id(topic['UI'] + qualifier['UI']),
                     "scheme": "MeSH",
                     "subject": topic['MH'] + "/" + qualifier['SH']
                 }
 
 
+@dataclass
+class Subject:
+    """Minimalist subject data."""
+
+    id: str
+    label: str
+
+
+@dataclass
+class Analysis:
+    """Holds results of analyzing a src subject."""
+
+    id: str
+    seen: bool = False  # present in dst subjects
+    relabelled: str = ""  # new label
+    replaced: str = ""  # id of replacing subject
+
+
+class MeSHSubjectDeltasConverter:
+    """Converter between subjects in DB + new subject file to operations."""
+
+    def __init__(self, src_subjects, dst_subjects, replacements):
+        """Constructor.
+
+        :param src_subjects: MeSH subjects found in the instance
+        :type src_subjects: Iterable[Subject]
+        :param dst_subjects: new subjects reader
+        :type dst_subjects: Iterable[Subject]
+        :param replacements: mapping of src labels to dst labels
+        :type replacements: dict[str, str]
+        """
+        self.src_subjects = src_subjects
+        self.dst_subjects = dst_subjects
+        self.replacements = replacements
+
+        # Will be filled out by _analyze_src(), _analyze_dst() and
+        # _analyze_replace()
+        self._id_to_analysis = {}
+        self._label_to_analysis = {}
+        self._additions = []
+
+        self.scheme = "MeSH"
+
+    def convert(self):
```

Comment:
This is the main important algorithm for generating delta operations.

The general idea is to load up the src subjects (coming from the database) in a convenient datastructure. Then iterate over the dst subjects (coming from the new MeSH file for the considered year) and see if they are new, renames or same. Then iterate over src subjects not seen in dst to check if they have been replaced or just removed. And finally, generate operations that summarize the above work.

Honestly this is the 3rd rewrite as a lot of unforeseen cases kept cropping up. In particular, it was discovered that a subject's qualifiers may change between years (added ones or removed ones). Please doublecheck (especially when looking at tests) if most cases are handled. 

Last thing to consider in your review is the question of scale/performance. I have not run this on a full dataset yet, so that will be the most informative. The approach does store in memory all previous subjects (there can be ~500K), but I don't know of another way (well... maybe another way would be to create a temporary DB table and leverage SQL queries to do large comparisons but that's also a lot of effort)

Reply:
I ran this on my local instance.

It took 33s on avg (after application loaded) to generate all the deltas (15 859 for transition to 2024 list of subjects including qualifiers - that's ~650K subjects). Memory didn't crash which is good. I may do a little more memory analysis, but this seems ok for now.

Reply:
Last update on this. After doing some profiling, it is indeed loading the src subjects from DB that take the most memory. The initial peak memory usage was at 2.3GiB. I was able to shave off close to 1GiB and reduce time by 62% in https://github.com/galterlibrary/galter-subjects-utils/pull/12 by extracting only the json from the DB. The peak memory usage is now in the order of 1.4GiB. 

Before:

![image](https://github.com/galterlibrary/galter-subjects-utils/assets/1932651/8a17f2bd-58a8-4c2b-99b0-ab90c8d5f7d5)

After:

![image](https://github.com/galterlibrary/galter-subjects-utils/assets/1932651/2ee00218-d88e-4e1b-af49-2a807177962d)

Reply:
Oh wow, nicely done Guillaume! I just happened to notice this while doing something else after the page refreshed.